framework/aster-frame/src/mm/io.rs

```diff
@@ -168,28 +176,133 @@ impl_vmio_pointer!(&mut T, "(**self)");
 impl_vmio_pointer!(Box<T>, "(**self)");
 impl_vmio_pointer!(Arc<T>, "(**self)");
 
-/// VmReader is a reader for reading data from a contiguous range of memory.
-pub struct VmReader<'a> {
+mod private {
+    pub trait Sealed {}
+}
+
+/// A marker structure used for [`VmReader`] and [`VmWriter`],
+/// representing their operated memory scope is in user space.
+pub struct UserSpace;
+
+/// A marker structure used for [`VmReader`] and [`VmWriter`],
+/// representing their operated memory scope is in kernel space.
+pub struct KernelSpace;
+
+impl private::Sealed for UserSpace {}
+impl private::Sealed for KernelSpace {}
+
+/// Copies `len` bytes from `src` to `dst`.
+/// Moves `src` and `dst` to the end of copied range.
+///
+/// Returns the number of successfully copied bytes.
+///
+/// # Safety
+///
+/// Mappings of virtual memory range [`src`..`src` + len] and [`dst`..`dst` + len]
+/// must be pinned. If one of the memory represent typed memory, these two virtual
+/// memory ranges and their corresponding physical pages should _not_ overlap.
+///
+/// Operation on typed memory may be safe only if it is plain-old-data. Otherwise
+/// the safety requirements of [`core::ptr::copy`] should also be considered.
+unsafe fn copy_with_pinned_memmap(src: &mut *const u8, dst: &mut *mut u8, len: usize) -> usize {
+    core::ptr::copy(*src, *dst, len);
```

Comment:
Actually, I wonder if this `ptr::copy` could (theoretically) lead to soundness problems, since the safety condition of `ptr::copy` [requires no concurrent access](https://doc.rust-lang.org/std/ptr/index.html#safety), which is not true in our case.

The memory copy primitives used in `VmReader`/`VmWriter`, even when `Target = KernelSpace`, are still very different from copying typed memory for the following two reasons:
 - We can clone a `Frame` and send it to another task, then create concurrent `VmReader`s and `VmWriter`s, which is not possible for normal Rust types because the mutable reference must be exclusive.
 - We can map a `Frame` to userspace via safe APIs, then we can guarantee literally nothing about how user programs operate on its userspace memory. They can access the shared page via atomic or non-atomic loads/stores.

I've checked what the Rust community thinks about this issue, and there are lots of efforts toward standard Rust APIs that can operate on memory memory shared with user programs (e.g., our `Frame`s) or memory shared with hardware (e.g., page tables and MMIOs), however, as of today, there is still no consensus on what those APIs should look like. If we are serious about this problem, the only answer is to accept the UB or develop our own APIs using inline assembly.

Historically, the C/C++ memory model [treats all data races as UBs](https://en.cppreference.com/w/cpp/language/multithread), which unfortunately makes certain data structures [impossible to implement without introducing UBs](https://github.com/rust-lang/unsafe-code-guidelines/blob/7a295dc32314830da58bdfd27bc186d9f4ae09ca/resources/deliberate-ub.md?plain=1#L8-L11). The Rust memory model pretty much inherits the C/C++ memory model, so it also [treats all data races as UBs](https://doc.rust-lang.org/reference/behavior-considered-undefined.html).

The LLVM memory model is a bit weaker than the C/C++ memory model, where racy memory loads are not UBs. Instead, such a memory load [returns `undef`](https://llvm.org/docs/Atomics.html#notatomic). However, `undef` is still not a well-defined value, as using `undef` can [easily trigger a UB](https://llvm.org/docs/LangRef.html#undefined-values). For example, if `x = undef`, the value of `x xor x` is not necessarily zero and the value of `x == x` is not necessarily true.

There is [a long discussion in the Rust community](https://github.com/rust-lang/unsafe-code-guidelines/issues/152), where one of the main questions discussed is whether `ptr::read_volatile`/`ptr::write_volatile` can access the memory region that another thread or some untrusted code (e.g. the userspace program or the hardware) can concurrently access. But as of today, they do not seem to have reached a consensus ([summary](https://github.com/rust-lang/unsafe-code-guidelines/issues/152#issuecomment-1261481330)). There seems to be evidence that the LLVM community is [against it](https://github.com/rust-lang/miri/issues/2807#issuecomment-1460657610). So someone has tried to propose [a new API for this purpose](https://github.com/rust-lang/unsafe-code-guidelines/issues/321), but unfortunately it is still in a pre-pre-RFC stage (after two years since the initial proposal), so I do not expect an implementation in the foreseeable future.

Well, this might be worth a separate issue (and outside the scope of this PR). To the best of my knowledge, if we prefer to avoid inline assembly, the best available solution is to use the volatile version, i.e. [intrinsics::volatile_copy_memory](https://doc.rust-lang.org/std/intrinsics/fn.volatile_copy_memory.html), to operate on the memory that is potentially shared with the user program or hardware. I believe this is still UB given the C/C++/Rust memory model, but this should be the best we can do.

Reply:
> I've checked what the Rust community thinks about this issue, and there are lots of efforts toward standard Rust APIs that can operate on memory memory shared with user programs (e.g., our Frames) or memory shared with hardware (e.g., page tables and MMIOs), however, as of today, there is still no consensus on what those APIs should look like. If we are serious about this problem, the only answer is to accept the UB or develop our own APIs using inline assembly.

Most of their discussions are irrelevant for our case because we have the powerful combination of untyped memory and plain-old data (POD). Data races are allowed for untyped memory because (1) no type-safe Rust objects are stored on untyped memory and (2) only POD objects can be read from and written into untyped memory. I think our solution is sound and powerful enough for our usage.

Reply:
**TL;DR:** I _strongly_ disagree, see the explanation below for details.

> Most of their discussions are irrelevant for our case because we have the powerful combination of untyped memory and plain-old data (POD).

They also talk about untyped memory. Any memory that can be shared with the hardware or with the user program _cannot_ be typed, otherwise this is a _strict_ violation of Rust's memory alias rules. No one is trying to fix this, it is unfixable, neither we nor the discussion in the Rust community.

> Data races are allowed for untyped memory because (1) no type-safe Rust objects are stored on untyped memory and (2) only POD objects can be read from and written into untyped memory. 

The problem is that whether the memory is typed or not, when we try to access the memory with `ptr::copy`, we are telling the compiler that data races cannot happen here, so the compiler can reason about possible optimizations _assuming no data races_. Those optimizations may be completely invalid if data races ever occur. This really has little to do with whether the memory is typed or not. 

> I think our solution is sound and powerful enough for our usage.

You must argue that _de facto_ (since you're violating the language/compiler's memory model) compiler optimizations will not miscompile the program even if the data race rule is violated. However, I think this argument _is_ wrong, here is a possible example:
```rust
    let index = vm_reader.read::<u32>(); // (1)
    if index >= vec.len() {
        return Err(());
    }

    // do a lot of unrelated things, but without any synchronization primitives

    let data = vec[index]; // (2)
    // do more things with `data`
```
In the above code snippet, the `vm_reader` reads the index of a region of memory shared with a user program, and then checks to see if the index is within the boundary of a `Vec`. After doing a lot of unrelated things, it tries to access the `Vec` using the index.

Since "a log of unrelated things" does not contain any synchronization primitives, and the compiler assumes that the data races cannot happen, the compiler may choose to discard the originally loaded value at (1), insert a spurious read at (2), and think that the spurious read will load the same value (that was checked in the bound before). However, if the value is changed in the middle by the user program, this will cause an out-of-bound access.

Similar bugs in C/C++ programs where the missing "volatile" keyword causes time-of-check time-of-use (ToCToU) races, eventually leading to serious security problems, such as [PWN2OWN](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1152r0.html#biblio-pwn2own) and [XENXSA155](https://insinuator.net/2015/12/xen-xsa-155-double-fetches-in-paravirtualized-devices/) (credit: I found them [here](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1152r0.html#extern)).

So we need at least the semantics of `volatile`. Using `volatile` with data races are still UBs, as documented and emphasized in the [Rust documentation](https://doc.rust-lang.org/std/ptr/fn.read_volatile.html#safety). However, using the `volatile` keyword is the _de facto_ way to handle similar problems, so it should be reliable enough (otherwise I think the Linux kernel will break too).

Here is another perspective from the LLVM memory model: The LLVM LangRef states that [a non-volatile load returns `undef` when data races occur](https://llvm.org/docs/LangRef.html#memory-model-for-concurrent-operations), which is the cause of the above problem. The memory location stores `undef`, so an erroneous read can see a completely different value, causing undefined behavior. It says [what a violate load returns is target-dependent](https://llvm.org/docs/LangRef.html#memory-model-for-concurrent-operations) (even if data races occur), and what we really need here is to guarantee that a volatile load will _never_ return `undef`. Unfortunately, the LLVM community [refuses to make this guarantee](https://github.com/llvm/llvm-project/issues/41780#issuecomment-981015344). But it should be the _de facto_ guarantee, at least among all the backend targets we can care about. Again, the Linux kernel will break if this is violated. If we don't believe this, we should use inline assembly instead.